File: 1 - metodosBusca/pacoteBusca_JomiHubner/src/exemplos/Estado8Puzzle.java
package exemplos;

import java.util.LinkedList;
import java.util.List;

import busca.Antecessor;
import busca.BuscaLargura;
import busca.Estado;
import busca.Heuristica;
import busca.MostraStatusConsole;
import busca.Nodo;

/**
 * Representa um estado do mundo: onde estao os numeros
 */
public class Estado8Puzzle implements Estado, Heuristica, Antecessor {
    
    public String getDescricao() {
        return
        "Este problema consiste em posicionar 8\n" +
        "numeros em um tabuleiro 3x3 na seguinte disposcao:\n"+
        "1 2 3\n"+
        "8   4\n"+
        "7 6 5\n"+
        "sendo que o espaco pode se mover.\n"+
        "(bom candidato a busca bi-direcional)\n";
    }
    
    public static final short tam = 3;
    
    int[][] tabuleiro = new int[tam][tam];
    int colBranco = -1;
    int linBranco = -1;
    
    /**
     *  cria um estado inicial (aleatorio)
     */
    public Estado8Puzzle() {
        for (int r=0;r< (tam*tam);r++) {
            // tenta ata achar uma posicao livre
            int l = Math.round( (float)(Math.random()*(tam-1)) );
            int c = Math.round( (float)(Math.random()*(tam-1)) );
            while (tabuleiro[l][c] != 0) {
                l = Math.round( (float)(Math.random()*(tam-1)) );
                c = Math.round( (float)(Math.random()*(tam-1)) );
            }
            tabuleiro[l][c] = r;
        }
        setPosBranco();
    }
    
    /**
     * cria um estado igual a outro
     */
    Estado8Puzzle(int[][] p) {
        for (int l=0;l<tam;l++) {
            //System.arraycopy(p,0,tabuleiro,0,tam);
            for (int c=0;c<tam;c++) {
                tabuleiro[l][c] = p[l][c];
            }
            
        }
    }
    
    void setPosBranco() {
        for (int l=0;l<tam;l++) {
            for (int c=0;c<tam;c++) {
                if (tabuleiro[l][c] == 0) {
                    colBranco = c;
                    linBranco = l;
                }
            }
        }
    }
    
    
    /**
     * verifica se um estado e igual a outro
     */
    public boolean equals(Object o) {
    	if (o instanceof Estado8Puzzle) {
            Estado8Puzzle e = (Estado8Puzzle)o;
            for (int l=0;l<tam;l++) {
                for (int c=0;c<tam;c++) {
                    if (tabuleiro[l][c] != e.tabuleiro[l][c]) {
                        return false;
                    }
                }
            }
            return true;
        }
        return false;
    }
    
    public int hashCode() {
        return toString().hashCode();
    }

    /**
     * verifica se o estado e meta
     */
    public boolean ehMeta() {
        return this.equals(estadoMeta);
    }
    
    
    /**
     * Heuristica: calcula a quantidade de numeros fora do lugar
     */
    public int h() {
        return h2() + h3();
    }
    
    /**
     * Heuristica: calcula a quantidade de numeros fora do lugar
     */
    public int h1() {
        int fora = 0;
        
        if (tabuleiro[0][0] != 1) fora++;
        if (tabuleiro[0][1] != 2) fora++;
        if (tabuleiro[0][2] != 3) fora++;
        if (tabuleiro[1][0] != 8) fora++;
        if (tabuleiro[1][1] != 0) fora++;
        if (tabuleiro[1][2] != 4) fora++;
        if (tabuleiro[2][0] != 7) fora++;
        if (tabuleiro[2][1] != 6) fora++;
        if (tabuleiro[2][2] != 5) fora++;
        
        return fora;
    }
    
    
    /**
     * Heuristica: calcula a distancia de cada numero ate seu lugar
     */
    public int h2() {
        int fora = 0;
        
        for (int n=0; n<(tam*tam); n++) {
            int l = getLinNro(n);
            int c = getColNro(n);
            int lMeta = estadoMeta.getLinNro(n);
            int cMeta = estadoMeta.getColNro(n);
            fora += Math.abs(l - lMeta);
            fora += Math.abs(c - cMeta);
        }
        return fora;
    }
    
    
    /**
     * Heuristica: conta os numeros fora da sequencia circular
     */
    public int h3() {
        int fora = 0;
        
        for (int n=1; n<(tam*tam); n++) {
            int l = getLinNro(n);
            int c = getColNro(n);
            int lAnt = 0;
            if (l==0 && c==0) lAnt = 1;
            else if (l==0 && c==1) lAnt = 0;
            else if (l==0 && c==2) lAnt = 0;
            else if (l==1 && c==0) lAnt = 2;
            else if (l==1 && c==2) lAnt = 0;
            else if (l==2 && c==0) lAnt = 2;
            else if (l==2 && c==1) lAnt = 2;
            else if (l==2 && c==2) lAnt = 1;
            
            int cAnt = 0;
            if (l==0 && c==0) cAnt = 0;
            else if (l==0 && c==1) cAnt = 0;
            else if (l==0 && c==2) cAnt = 1;
            else if (l==1 && c==0) cAnt = 0;
            else if (l==1 && c==2) cAnt = 2;
            else if (l==2 && c==0) cAnt = 1;
            else if (l==2 && c==1) cAnt = 2;
            else if (l==2 && c==2) cAnt = 2;
            
            int nroAnt = tabuleiro[lAnt][cAnt];
            
            if (n == 1) {
                if (nroAnt != 8) {
                    fora += 1;
                }
            } else if (nroAnt+1 != n) {
                fora += 1;
            }
        }
        return fora;
    }
    
    /** retorna a coluna de um numero */
    int getColNro(int n) {
        for (int l=0;l<tam;l++) {
            for (int c=0;c<tam;c++) {
                if (tabuleiro[l][c] == n) {
                    return c;
                }
            }
        }
        return -1;
    }
    
    /** retorna a linha de um numero */
    int getLinNro(int n) {
        for (int l=0;l<tam;l++) {
            for (int c=0;c<tam;c++) {
                if (tabuleiro[l][c] == n) {
                    return l;
                }
            }
        }
        return -1;
    }
    
    
    /**
     * gera uma lista de sucessores do nodo.
     */
    public List<Estado> sucessores() {
        List<Estado> suc = new LinkedList<Estado>(); // a lista de sucessores
        
        // pra cima
        if (linBranco > 0) {
            Estado8Puzzle novo = new Estado8Puzzle(tabuleiro);
            novo.tabuleiro[linBranco-1][colBranco] = 0;
            novo.tabuleiro[linBranco][colBranco] = tabuleiro[linBranco-1][colBranco];
            novo.linBranco = linBranco-1;
            novo.colBranco = colBranco;
            suc.add(novo);
        }
        
        // pra baixo
        if (linBranco < tam-1) {
            Estado8Puzzle novo = new Estado8Puzzle(tabuleiro);
            novo.tabuleiro[linBranco+1][colBranco] = 0;
            novo.tabuleiro[linBranco][colBranco] = tabuleiro[linBranco+1][colBranco];
            novo.linBranco = linBranco+1;
            novo.colBranco = colBranco;
            suc.add(novo);
        }
        
        // pra esq
        if (colBranco > 0) {
            Estado8Puzzle novo = new Estado8Puzzle(tabuleiro);
            novo.tabuleiro[linBranco][colBranco-1] = 0;
            novo.tabuleiro[linBranco][colBranco] = tabuleiro[linBranco][colBranco-1];
            novo.linBranco = linBranco;
            novo.colBranco = colBranco-1;
            suc.add(novo);
        }
        
        // pra dir
        if (colBranco < tam-1) {
            Estado8Puzzle novo = new Estado8Puzzle(tabuleiro);
            novo.tabuleiro[linBranco][colBranco+1] = 0;
            novo.tabuleiro[linBranco][colBranco] = tabuleiro[linBranco][colBranco+1];
            novo.linBranco = linBranco;
            novo.colBranco = colBranco+1;
            suc.add(novo);
        }
        
        return suc;
    }
    
    public List<Estado> antecessores() {
        return sucessores();
    }
    
    public boolean temSolucao() {
        // se um nro par de trocar levar o estado corrente para a meta, 
    	// entao tem solucao
        Estado8Puzzle meta = estadoMeta;
        
        Estado8Puzzle t = new Estado8Puzzle(tabuleiro);
        
        // move o branco para o centro
        int cb = t.getColNro(0);
        int lb = t.getLinNro(0);
        
        // move linha
        while ( lb < meta.getLinNro(0)) {
            troca( t.tabuleiro, lb, cb, lb+1, cb);
            lb++;
        }
        while ( lb > meta.getLinNro(0)) {
            troca( t.tabuleiro, lb, cb, lb-1, cb);
            lb--;
        }
        
        // move coluna
        while ( cb < meta.getColNro(0)) {
            troca( t.tabuleiro, lb, cb, lb, cb+1);
            cb++;
        }
        while ( cb > meta.getColNro(0)) {
            troca( t.tabuleiro, lb, cb, lb, cb-1);
            cb--;
        }
        
        
        int nroTrocas = 0;
        for (int l = 0; l < tam; l++) {
            for (int c = 0; c < tam; c++) {
                // se meta for diferente de tab, troca
                if (t.tabuleiro[l][c] != meta.tabuleiro[l][c]) {
                    int vlOk = meta.tabuleiro[l][c];
                    //Estado8Puzzle t = new Estado8Puzzle(tab);
                    troca(t.tabuleiro, l, c, t.getLinNro(vlOk), t.getColNro(vlOk));
                    nroTrocas++;
                }
            }
        }
        
        return nroTrocas % 2 == 0;
        
    }
    
    private void troca(int[][] tab, int l1, int c1, int l2, int c2) {
        int bak = tab[l2][c2];
        tab[l2][c2] = tab[l1][c1];
        tab[l1][c1] = bak;
    }
    
    private String toStringCache = null;
    public String toString() {
        if (toStringCache == null) {
            StringBuffer r = new StringBuffer("\n");
            for (int i=0;i<tam;i++) {
                for (int j=0;j<tam;j++) {
                    r.append(tabuleiro[i][j]);
                    if (j+1<tam) {
                        r.append(" ");
                    }
                }
                if (i+1<tam) {
                    r.append("\n");
                }
            }
            toStringCache = r + "\n";
        }
        return toStringCache;
    }
    
    
    /**
     * Custo para geracao de um estado
     */
    public int custo() {
        return 1;
    }
    
    
    static public Estado8Puzzle getEstadoFacil() {
        Estado8Puzzle e8 = new Estado8Puzzle(new int[][] {{8,1,3},{0,7,2},{6,5,4}});
        e8.setPosBranco();
        return e8;
    }
    
    static public Estado8Puzzle getEstadoDificil() {
        Estado8Puzzle e8 = new Estado8Puzzle(new int[][] {{7,8,6},{2,3,5},{1,4,0}} );
        e8.setPosBranco();
        return e8;
    }
    
    static public Estado8Puzzle getEstadoMuitoDificil() {
        //Estado8Puzzle e8 = new Estado8Puzzle(new int[][] {{3,4,8},{1,2,5},{7,0,6}} );
        Estado8Puzzle e8 = new Estado8Puzzle(new int[][] {{2,3,4},{1,0,5},{8,7,6}} );
        e8.setPosBranco();
        return e8;
    }
    
    
    
    public static Estado8Puzzle getEstadoMeta() {
        return estadoMeta;
    }
    
    private final static Estado8Puzzle estadoMeta = setEstadoMeta();
    private static Estado8Puzzle setEstadoMeta() {
        Estado8Puzzle e8 = new Estado8Puzzle(new int[][] {{1,2,3},{8,0,4},{7,6,5}});
        e8.setPosBranco();
        return e8;
    }
    
    public static void main(String[] a) {
        //Estado8Puzzle e8 = getEstadoFacil();
        Estado8Puzzle e8 = getEstadoDificil();
        //Estado8Puzzle e8 = getEstadoMuitoDificil();
        System.out.println("estado inicial (h="+((Heuristica)e8).h()+") ="+e8);
        
        if (! e8.temSolucao()) {
            System.out.println(e8+"nao tem solucao!");
            return;
        }
        
        //Nodo s = new AEstrela().busca(e8);
        Nodo s = new BuscaLargura(new MostraStatusConsole()).busca(e8);
        //Nodo s = new BuscaIterativo(new MostraStatusConsole()).busca(e8);
        //Nodo s = new BuscaProfundidade(25,new MostraStatusConsole()).busca(e8);
        //Nodo s = new BuscaBidirecional(new MostraStatusConsole()).busca(e8, getEstadoMeta());
        if (s != null) {
            System.out.println("solucao ("+s.getProfundidade()+")= "+s.montaCaminho());
        }        
    }
}

File: 1 - metodosBusca/python/app/exemplos/estado_8_puzzle.py
import random

from app.core.busca import (
    Antecessor,
    BuscaLargura,
    Estado,
    Heuristica,
    MostraStatusConsole,
)


class Estado8Puzzle(Estado, Heuristica, Antecessor):
    tam = 3
    estado_meta = None

    def __init__(self, p=None):
        self.tabuleiro = [[0] * self.tam for _ in range(self.tam)]
        self.col_branco = -1
        self.lin_branco = -1

        if p is None:
            for r in range(self.tam * self.tam):
                l = random.randint(0, self.tam - 1)
                c = random.randint(0, self.tam - 1)
                while self.tabuleiro[l][c] != 0:
                    l = random.randint(0, self.tam - 1)
                    c = random.randint(0, self.tam - 1)
                self.tabuleiro[l][c] = r
            self.set_pos_branco()
        else:
            for l in range(self.tam):
                for c in range(self.tam):
                    self.tabuleiro[l][c] = p[l][c]
            self.set_pos_branco()

    def get_descricao(self):
        return (
            "Este problema consiste em posicionar 8\n"
            "numeros em um tabuleiro 3x3 na seguinte disposcao:\n"
            "1 2 3\n"
            "8   4\n"
            "7 6 5\n"
            "sendo que o espaco pode se mover.\n"
            "(bom candidato a busca bi-direcional)\n"
        )

    def set_pos_branco(self):
        for l in range(self.tam):
            for c in range(self.tam):
                if self.tabuleiro[l][c] == 0:
                    self.col_branco = c
                    self.lin_branco = l

    def __eq__(self, other):
        if isinstance(other, Estado8Puzzle):
            return self.tabuleiro == other.tabuleiro
        return False

    def __hash__(self):
        return hash(str(self))

    def eh_meta(self):
        return self == self.estado_meta

    def h(self):
        return self.h2() + self.h3()

    def h1(self):
        fora = 0
        if self.tabuleiro[0][0] != 1: fora += 1
        if self.tabuleiro[0][1] != 2: fora += 1
        if self.tabuleiro[0][2] != 3: fora += 1
        if self.tabuleiro[1][0] != 8: fora += 1
        if self.tabuleiro[1][1] != 0: fora += 1
        if self.tabuleiro[1][2] != 4: fora += 1
        if self.tabuleiro[2][0] != 7: fora += 1
        if self.tabuleiro[2][1] != 6: fora += 1
        if self.tabuleiro[2][2] != 5: fora += 1
        return fora

    def h2(self):
        fora = 0
        for n in range(self.tam * self.tam):
            l = self.get_lin_nro(n)
            c = self.get_col_nro(n)
            l_meta = self.estado_meta.get_lin_nro(n)
            c_meta = self.estado_meta.get_col_nro(n)
            fora += abs(l - l_meta) + abs(c - c_meta)
        return fora

    def h3(self):
        fora = 0
        for n in range(1, self.tam * self.tam):
            l = self.get_lin_nro(n)
            c = self.get_col_nro(n)
            l_ant, c_ant = self.get_antecessor_pos(l, c)
            nro_ant = self.tabuleiro[l_ant][c_ant]
            if n == 1:
                if nro_ant != 8:
                    fora += 1
            elif nro_ant + 1 != n:
                fora += 1
        return fora

    def get_col_nro(self, n):
        for l in range(self.tam):
            for c in range(self.tam):
                if self.tabuleiro[l][c] == n:
                    return c
        return -1

    def get_lin_nro(self, n):
        for l in range(self.tam):
            for c in range(self.tam):
                if self.tabuleiro[l][c] == n:
                    return l
        return -1

    def get_antecessor_pos(self, l, c):
        if l == 0 and c == 0: return 1, 0
        if l == 0 and c == 1: return 0, 0
        if l == 0 and c == 2: return 0, 1
        if l == 1 and c == 0: return 2, 0
        if l == 1 and c == 2: return 0, 2
        if l == 2 and c == 0: return 2, 1
        if l == 2 and c == 1: return 2, 2
        if l == 2 and c == 2: return 1, 2
        return 0, 0

    def sucessores(self):
        suc = []
        if self.lin_branco > 0:
            novo = Estado8Puzzle(self.tabuleiro)
            novo.tabuleiro[self.lin_branco - 1][self.col_branco] = 0
            novo.tabuleiro[self.lin_branco][self.col_branco] = self.tabuleiro[self.lin_branco - 1][self.col_branco]
            novo.lin_branco = self.lin_branco - 1
            novo.col_branco = self.col_branco
            suc.append(novo)
        if self.lin_branco < self.tam - 1:
            novo = Estado8Puzzle(self.tabuleiro)
            novo.tabuleiro[self.lin_branco + 1][self.col_branco] = 0
            novo.tabuleiro[self.lin_branco][self.col_branco] = self.tabuleiro[self.lin_branco + 1][self.col_branco]
            novo.lin_branco = self.lin_branco + 1
            novo.col_branco = self.col_branco
            suc.append(novo)
        if self.col_branco > 0:
            novo = Estado8Puzzle(self.tabuleiro)
            novo.tabuleiro[self.lin_branco][self.col_branco - 1] = 0
            novo.tabuleiro[self.lin_branco][self.col_branco] = self.tabuleiro[self.lin_branco][self.col_branco - 1]
            novo.lin_branco = self.lin_branco
            novo.col_branco = self.col_branco - 1
            suc.append(novo)
        if self.col_branco < self.tam - 1:
            novo = Estado8Puzzle(self.tabuleiro)
            novo.tabuleiro[self.lin_branco][self.col_branco + 1] = 0
            novo.tabuleiro[self.lin_branco][self.col_branco] = self.tabuleiro[self.lin_branco][self.col_branco + 1]
            novo.lin_branco = self.lin_branco
            novo.col_branco = self.col_branco + 1
            suc.append(novo)
        return suc

    def antecessores(self):
        return self.sucessores()

    def tem_solucao(self):
        meta = self.estado_meta
        t = Estado8Puzzle(self.tabuleiro)
        cb = t.get_col_nro(0)
        lb = t.get_lin_nro(0)
        while lb < meta.get_lin_nro(0):
            self.troca(t.tabuleiro, lb, cb, lb + 1, cb)
            lb += 1
        while lb > meta.get_lin_nro(0):
            self.troca(t.tabuleiro, lb, cb, lb - 1, cb)
            lb -= 1
        while cb < meta.get_col_nro(0):
            self.troca(t.tabuleiro, lb, cb, lb, cb + 1)
            cb += 1
        while cb > meta.get_col_nro(0):
            self.troca(t.tabuleiro, lb, cb, lb, cb - 1)
            cb -= 1
        nro_trocas = 0
        for l in range(self.tam):
            for c in range(self.tam):
                if t.tabuleiro[l][c] != meta.tabuleiro[l][c]:
                    vl_ok = meta.tabuleiro[l][c]
                    self.troca(t.tabuleiro, l, c, t.get_lin_nro(vl_ok), t.get_col_nro(vl_ok))
                    nro_trocas += 1
        return nro_trocas % 2 == 0

    def troca(self, tab, l1, c1, l2, c2):
        tab[l1][c1], tab[l2][c2] = tab[l2][c2], tab[l1][c1]

    def __str__(self):
        if not hasattr(self, 'to_string_cache'):
            r = "\n"
            for i in range(self.tam):
                r += " ".join(map(str, self.tabuleiro[i])) + "\n"
            self.to_string_cache = r
        return self.to_string_cache

    def custo(self):
        return 1

    @staticmethod
    def get_estado_facil():
        e8 = Estado8Puzzle([[8, 1, 3], [0, 7, 2], [6, 5, 4]])
        e8.set_pos_branco()
        return e8

    @staticmethod
    def get_estado_dificil():
        e8 = Estado8Puzzle([[7, 8, 6], [2, 3, 5], [1, 4, 0]])
        e8.set_pos_branco()
        return e8

    @staticmethod
    def get_estado_muito_dificil():
        e8 = Estado8Puzzle([[2, 3, 4], [1, 0, 5], [8, 7, 6]])
        e8.set_pos_branco()
        return e8

    def get_estado_meta(self):
        return self.estado_meta
    
    def set_estado_meta(self):
        self.estado_meta = Estado8Puzzle([[1, 2, 3], [8, 0, 4], [7, 6, 5]])
        self.estado_meta.set_pos_branco()

        

if __name__ == "__main__":
    e8 = Estado8Puzzle.get_estado_dificil()
    e8.set_estado_meta()
    print(f"estado inicial (h={e8.h()}) = {e8}")

    if not e8.tem_solucao():
        print(f"{e8} nao tem solucao!")
    else:
        s = BuscaLargura(MostraStatusConsole()).busca(e8)
        if s is not None:
            print(f"solucao ({s.get_profundidade()})= {s.monta_caminho()}")
